### PR TITLE
OBPIH-6536 [backend] Create partial invoice for prepay PO

### PR DESCRIPTION
**Pull request by alannadolny**
### :sparkles: Description of Change

> *A concise summary of what is being changed. Please provide enough context for reviewers to be able to understand the change and why it is necessary.*

**Link to GitHub issue or Jira ticket:** https://pihemr.atlassian.net/browse/OBPIH-6536

**Description:**

:x: 

---
### :camera: Screenshots & Recordings (optional)
:x:

---
### :chart_with_upwards_trend: Test Plan

> *We require that all code changes come paired with a method of testing them. Please select which of the following testing approaches you've included with this change:*

- [ ] Frontend automation tests (unit)
- [x] Backend automation tests ([unit](https://pihemr.atlassian.net/wiki/spaces/OB/pages/3013869579/Backend+Unit+Testing+Standards), [API](https://pihemr.atlassian.net/wiki/spaces/OB/pages/3013738522/Backend+Integration+Testing+Standards), smoke)
- [ ] End-to-end tests ([Playwright](https://pihemr.atlassian.net/wiki/spaces/OB/pages/2942828546/Knowledge+Transfer+on+automated+testing+with+Playwright))
- [x] Manual tests (please describe below)
- [ ] Not Applicable

**Description of test plan (if applicable):**
Check whether only not invoiced items/adjustments are visible on the Create Invoice page

---
### :white_check_mark: Quality Checks

> *Please confirm and check each of the following to ensure that your change conforms to the coding standards of the project:*

- [x] The pull request title is prefixed with the issue/ticket number (Ex: `[OBS-123]` for Jira, `[#0000]` for GitHub, or `[OBS-123, OBPIH-123]` if there are multiple), or with `[N/A]` if not applicable
- [x] The pull request description has enough information for someone without context to be able to understand the change and why it is needed
- [x] The change has tests that prove the issue is fixed / the feature works as intended

2nd condition is met - the context is available in the ticket

### CODE REVIEW

File: grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy [hunk 1]
```diff
@@ -107,7 +108,7 @@ class InvoiceController {
             redirect(action: "list")
         }
 
-        Invoice invoice = invoiceService.generateInvoice(order)
+        Invoice invoice = prepaymentInvoiceService.generateInvoice(order)
```
Comment (EWaterman) [inline grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy:111]: What method is getting called when generating an invoice for a PO that wasn't pre-paid? If this API is used in non-prepay related flows we don't want to have it use the prepaymentInvoiceService

File: src/test/groovy/unit/org/pih/warehouse/invoice/InvoiceServiceSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.pih.warehouse.product.Product
+import org.pih.warehouse.shipping.ShipmentItem
+import spock.lang.Specification
+import spock.lang.Unroll
+
+@Unroll
+class InvoiceServiceSpec extends Specification implements ServiceUnitTest<InvoiceService>, DataTest {
+
+    @Override
+    Class<?>[] getDomainClassesToMock() {
```
Comment (EWaterman) [inline src/test/groovy/unit/org/pih/warehouse/invoice/InvoiceServiceSpec.groovy:18]: why is this needed? You can't just call mockDomains in setup?

File: grails-app/domain/org/pih/warehouse/shipping/ShipmentItem.groovy [hunk 3]
```diff
@@ -260,6 +278,14 @@ class ShipmentItem implements Comparable, Serializable {
         return invoiceItems?.sum { it.quantity ?: 0 } ?: 0
     }
 
+    Integer getQuantityToInvoice() {
+        return quantity - quantityInvoiced
```
Comment (EWaterman) [inline grails-app/domain/org/pih/warehouse/shipping/ShipmentItem.groovy:282]: What's your plan for eventually handling cancellations/returns?

quantityInvoiced loops through all the invoices and sums the quantity, which works fine for the case where there are no quantity adjustments but... if a shipment item has quantity 3 and we cancelled or returned the shipment, we generate an invoice with quantity 0. In that scenario quantityInvoiced for the shipment item is 0, and isInvoiceable will be true (which is incorrect). We'll need some way to note that even if we invoice for less than the quantity shipped for that shipment item, we've still "fully invoiced" it.

Side note: I remember chatting with Manon about whether we need to be able to split a single shipment item up into multiple invoices (Ex: we have a shipment item with quantity 3 which has an invoice item with quantity 1 and another invoice item with quantity 2) and I thiiiiink she said we don't need the ability to do that (probably worth confirming).

So maybe it'd just be a matter of changing the getQuantityInvoiced method to return 0 for shipment items without an invoice and shipmentItem.quantity for shipment items that do have an invoice. So we ignore the quantity in the invoiceItem for the purpose of determining if the shipment item has been fully invoiced. (Which we can do because we can safely assume that each shipmentItem will only ever have a single invoiceItem so if it was invoiced at all, it's done.)

I know the work to implement cancellations isn't covered by this ticket, I just want to make sure it'll eventually be able to work within this system.

File: grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy (hunk 1 above)
Comment (EWaterman) [reply grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy:111]: nvm I see the first line in the method checks that it has a prepayment invoice already.

Have you considered moving these methods to a PrepaymentInvoiceController since you're also doing that for the service methods?

File: grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import org.pih.warehouse.order.OrderAdjustment
+import org.pih.warehouse.order.OrderItem
+import org.pih.warehouse.order.OrderItemStatusCode
+import org.pih.warehouse.shipping.ShipmentItem
+import grails.gorm.transactions.Transactional
+
+@Transactional
+class PrepaymentInvoiceService {
+
+    InvoiceService invoiceService
```
Comment (EWaterman) [inline grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy:14]: Generally my preference would be to not have services call other services but instead move common code to shared util classes (Ex: InvoiceNumberGenerator, InvoiceItemGenerator (or whatever names)). It breaks the code up into more manageable units (easier testing too).

I don't expect you to to make any kind of change like that here though. Just noting down my thoughts.

File: grails-app/domain/org/pih/warehouse/order/Order.groovy [hunk 5]
```diff
@@ -462,6 +464,14 @@ class Order implements Serializable {
         return !hasInvoice && isPrepaymentRequired && hasActiveItemsOrAdjustments
     }
 
+    Set<OrderItem> getInvoiceableOrderItems() {
+        return orderItems.findAll { it.invoiceable }
+    }
+
+    Set<OrderAdjustment> getInvoiceableAdjustment() {
```
Comment (EWaterman) [inline grails-app/domain/org/pih/warehouse/order/Order.groovy:471]:
```suggestion
    Set<OrderAdjustment> getInvoiceableAdjustments() {
```

File: grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy (hunk 1 above)
Comment (alannadolny) [reply grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy:111]: I think it's not worth it, to be honest. Later we will add PrepaymentInvoiceApiController, so this one will have only one endpoint.
